pymc_experimental/model_transform/conditioning.py

```diff
@@ -0,0 +1,103 @@
+from typing import Any, Dict, List, Sequence, Union
+
+from pymc import Model
+from pymc.pytensorf import _replace_rvs_in_graphs
+from pytensor.tensor import TensorVariable
+
+from pymc_experimental.utils.model_fgraph import (
+    ModelFreeRV,
+    extract_dims,
+    fgraph_from_model,
+    model_from_fgraph,
+    model_named,
+    model_observed_rv,
+    toposort_replace,
+)
+
+
+def observe(model: Model, vars_to_observations: Dict[Union["str"], Any]) -> Model:
+    """Convert free RVS to observed RVs.
+
+    This function returns a new model.
+    """
+    vars_to_observations = {
+        model[var] if isinstance(var, str) else var: obs
+        for var, obs in vars_to_observations.items()
+    }
+
+    # Note: Since PyMC can infer logprob expressions we could also allow observing Deterministics
+    if any(var not in model.free_RVs for var in vars_to_observations):
+        raise ValueError(f"At least one var is not a free variable in the model")
+
+    fgraph, memo = fgraph_from_model(model)
+
+    replacements = {}
+    for var, obs in vars_to_observations.items():
+        model_free_rv = memo[var]
+
+        # Just a sanity check
+        assert isinstance(model_free_rv.owner.op, ModelFreeRV)
+        assert model_free_rv in fgraph.variables
+
+        rv, vv, *dims = model_free_rv.owner.inputs
+        model_obs_rv = model_observed_rv(rv, rv.type.filter_variable(obs), *dims)
```

Comment:
Could `obs` have also depended on some nodes from the original model graph? If yes, maybe we should check and replace with memo?

Reply:
Added a test for Minibatch observations. This should cover the whole range of "allowed" observed variables.